File: src/utils/Renderer.java
package utils;

import java.util.List;

import entidades.*;
import objetos.Objeto;

public class Renderer {
    public static final String RESET = "\u001B[0m";
    public static final String ROJO = "\u001B[31m";
    public static final String VERDE = "\u001B[32m";
    public static final String AMARILLO = "\u001B[33m";
    public static final String CYAN = "\u001B[36m";
    public static final String MAGENTA = "\u001B[35m";
    public static final String AZUL = "\u001B[34m";

    private static final String[] COLORES_ESTRELLA = {
        "\u001B[1;31m",
        "\u001B[1;33m",
        "\u001B[1;32m",
        "\u001B[1;36m",
        "\u001B[1;35m",
        "\u001B[1;37m",
    };

    public void mostrarTablero(GameBoard board, int turno) {
        int filas = board.getFilas();
        int columnas = board.getColumnas();
        StringBuilder sb = new StringBuilder(filas * columnas * 15);
        for (int i = 0; i < filas; i++) {
            for (int j = 0; j < columnas; j++) {
                Entidad e = board.getEntidad(i, j);
                if (e != null) {
                    char c = e.getSimbolo();
                    if (c == 'T') {
                        sb.append("\u001B[1;31m").append(" T ").append(RESET);
                    } else if (c == 'R') {
                        sb.append(AMARILLO).append(" \u00A4 ").append(RESET);
                    } else if (c == 'X') {
                        sb.append(ROJO).append(" # ").append(RESET);
                    } else if (c == 'A') {
                        if (e instanceof Aliado && ((Aliado) e).getTurnosInvencible() > 0) {
                            String color = COLORES_ESTRELLA[turno % COLORES_ESTRELLA.length];
                            sb.append(color).append(" o ").append(RESET);
                        } else {
                            sb.append(VERDE).append(" o ").append(RESET);
                        }
                    } else if (c == 'M') {
                        sb.append(AMARILLO).append("[=]").append(RESET);
                    } else {
                        sb.append(" . ");
                    }
                } else if (board.getObjeto(i, j) != null) {
                    char s = board.getObjeto(i, j).getSimbolo();
                    if (s == 'S') {
                        sb.append(CYAN).append(" S ").append(RESET);
                    } else if (s == 'W') {
                        sb.append(MAGENTA).append(" W ").append(RESET);
                    } else if (s == 'V') {
                        sb.append(AZUL).append(" V ").append(RESET);
                    } else if (s == '*') {
                        sb.append("\u001B[1;33m").append(" * ").append(RESET);
                    } else if (s == '+') {
                        sb.append("\u001B[1;32m").append(" + ").append(RESET);
                    } else {
                        sb.append(" ").append(s).append(" ");
                    }
                } else if (board.getTrampa(i, j) != null) {
                    sb.append("\u001B[37m").append(" ^ ").append(RESET);
                } else {
                    sb.append(" . ");
                }
            }
            sb.append("\033[K\n");
        }
        System.out.print(sb);
        System.out.flush();
    }

    public void mostrarHUD(GameBoard board, GameConfig config, int turno, long tiempoInicio,
                           int enemigosEliminados, int objetosRecogidos, int numAliados, int numEnemigos) {
        long tiempoMs = System.currentTimeMillis() - tiempoInicio;
        int segundosTotales = (int) (tiempoMs / 1000);
        int minutos = segundosTotales / 60;
        int segundos = segundosTotales % 60;
        String tiempo = String.format("%02d:%02d", minutos, segundos);

        String colorAliados = numAliados <= 2 ? ROJO : VERDE;

        int anchoHud = board.getColumnas() * 3;
        String lineaH = "\u2550".repeat(anchoHud - 2);

        String linea1 = String.format("  Turno: %d  \u2502  Tiempo: %s  \u2502  Velocidad: %dms",
                turno, tiempo, config.velocidadMs);
        String linea2 = String.format("  Aliados: %s%d/%d%s  \u2502  Enemigos: %s%d%s  \u2502  Eliminados: %s%d%s  \u2502  Objetos: %s%d%s",
                colorAliados, numAliados, config.numAliado, RESET,
                ROJO, numEnemigos, RESET,
                VERDE, enemigosEliminados, RESET,
                CYAN, objetosRecogidos, RESET);

        String leyenda = "  " + VERDE + "o" + RESET + "=Aliado  "
                + ROJO + "#" + RESET + "=Enemigo  "
                + "\u001B[1;31m" + "T" + RESET + "=Tanque  "
                + AMARILLO + "\u00A4" + RESET + "=R\u00E1pido  "
                + AMARILLO + "[=]" + RESET + "=Muro  "
                + "\u001B[37m" + "^" + RESET + "=Trampa  "
                + CYAN + "S" + RESET + "=Escudo  "
                + MAGENTA + "W" + RESET + "=Arma  "
                + "\u001B[1;33m" + "*" + RESET + "=Estrella  "
                + AZUL + "V" + RESET + "=Vel  "
                + "\u001B[1;32m" + "+" + RESET + "=Poci\u00F3n";

        System.out.println("\u2554" + lineaH + "\u2557\033[K");
        System.out.println("\u2551" + linea1 + espacios(anchoHud - 2 - longitudVisible(linea1)) + "\u2551\033[K");
        System.out.println("\u2551" + linea2 + espacios(anchoHud - 2 - longitudVisible(linea2)) + "\u2551\033[K");
        System.out.println("\u2551" + leyenda + espacios(anchoHud - 2 - longitudVisible(leyenda)) + "\u2551\033[K");
        System.out.println("\u255A" + lineaH + "\u255D\033[K");
    }

    public void mostrarResultado(String resultado) {
        switch (resultado) {
            case "empate":
                System.out.println(AMARILLO + " Empate! Ambos bandos han caido al mismo tiempo." + RESET);
                break;
            case "enemigos":
                System.out.println(ROJO + " Los enemigos han ganado! Todos los aliados han caido." + RESET);
                break;
            case "aliados":
                System.out.println(VERDE + " Los aliados han sobrevivido! Todos los enemigos han sido eliminados." + RESET);
                break;
        }
    }

    public void limpiarPantalla() {
        System.out.print("\033[H");
    }

    public void ocultarCursor() {
        System.out.print("\033[?25l\033[H\033[2J");
        System.out.flush();
    }

    public void restaurarCursor() {
        System.out.print("\033[?25h");
        System.out.flush();
    }

    public void mostrarEstadisticas(List<Entidad> todasEntidades, String resultado,
                                     int turnos, long tiempoInicio,
                                     int numAliadosInicial, int numEnemigosInicial) {
        int ancho = 46;
        String lineaH = "\u2550".repeat(ancho);

        // Calcular stats por faccion
        int danioAliadosInf = 0, danioAliadosRec = 0, killsAliados = 0, objRecogidos = 0;
        int aliadosVivos = 0;
        int danioEnemigosInf = 0, danioEnemigosRec = 0, killsEnemigos = 0;
        int enemigosVivos = 0;

        Entidad mvpAliado = null, mvpEnemigo = null;

        for (Entidad e : todasEntidades) {
            if (e instanceof Aliado) {
                Aliado a = (Aliado) e;
                danioAliadosInf += a.getDanioInfligido();
                danioAliadosRec += a.getDanioRecibido();
                killsAliados += a.getKills();
                objRecogidos += a.getObjetosRecogidosPersonal();
                if (a.estaVivo()) aliadosVivos++;
                if (mvpAliado == null || esMejorMvp(a, mvpAliado)) mvpAliado = a;
            } else if (e instanceof Enemigo) {
                danioEnemigosInf += e.getDanioInfligido();
                danioEnemigosRec += e.getDanioRecibido();
                killsEnemigos += e.getKills();
                if (e.estaVivo()) enemigosVivos++;
                if (mvpEnemigo == null || esMejorMvp(e, mvpEnemigo)) mvpEnemigo = e;
            }
        }

        long tiempoMs = System.currentTimeMillis() - tiempoInicio;
        int seg = (int) (tiempoMs / 1000);
        String tiempo = String.format("%02d:%02d", seg / 60, seg % 60);

        String textoResultado;
        switch (resultado) {
            case "aliados": textoResultado = VERDE + "Los aliados han ganado" + RESET; break;
            case "enemigos": textoResultado = ROJO + "Los enemigos han ganado" + RESET; break;
            default: textoResultado = AMARILLO + "Empate" + RESET; break;
        }

        StringBuilder sb = new StringBuilder();
        sb.append("\n");
        sb.append("\u2554").append(lineaH).append("\u2557\n");
        lineaCentrada(sb, ancho, "ESTADISTICAS DE PARTIDA");
        sb.append("\u2560").append(lineaH).append("\u2563\n");
        lineaIzq(sb, ancho, "  Resultado: " + textoResultado);
        lineaIzq(sb, ancho, "  Turnos: " + turnos + "  \u2502  Tiempo: " + tiempo);
        sb.append("\u2560").append(lineaH).append("\u2563\n");
        lineaCentrada(sb, ancho, VERDE + "ALIADOS" + RESET);
        lineaIzq(sb, ancho, "  Danio infligido total: " + danioAliadosInf);
        lineaIzq(sb, ancho, "  Danio recibido total: " + danioAliadosRec);
        lineaIzq(sb, ancho, "  Supervivientes: " + aliadosVivos + "/" + numAliadosInicial);
        lineaIzq(sb, ancho, "  Kills totales: " + killsAliados);
        lineaIzq(sb, ancho, "  Objetos recogidos: " + objRecogidos);
        sb.append("\u2560").append(lineaH).append("\u2563\n");
        lineaCentrada(sb, ancho, ROJO + "ENEMIGOS" + RESET);
        lineaIzq(sb, ancho, "  Danio infligido total: " + danioEnemigosInf);
        lineaIzq(sb, ancho, "  Danio recibido total: " + danioEnemigosRec);
        lineaIzq(sb, ancho, "  Supervivientes: " + enemigosVivos + "/" + numEnemigosInicial);
        lineaIzq(sb, ancho, "  Kills totales: " + killsEnemigos);
        sb.append("\u2560").append(lineaH).append("\u2563\n");

        if (mvpAliado != null) {
            lineaCentrada(sb, ancho, VERDE + "MVP ALIADO" + RESET);
            lineaIzq(sb, ancho, "  Aliado #" + mvpAliado.getId() + " \u2014 " + mvpAliado.getKills() + " kills, " + mvpAliado.getDanioInfligido() + " danio");
        }
        lineaIzq(sb, ancho, "");
        if (mvpEnemigo != null) {
            lineaCentrada(sb, ancho, ROJO + "MVP ENEMIGO" + RESET);
            lineaIzq(sb, ancho, "  Enemigo #" + mvpEnemigo.getId() + " \u2014 " + mvpEnemigo.getKills() + " kills, " + mvpEnemigo.getDanioInfligido() + " danio");
        }

        sb.append("\u255A").append(lineaH).append("\u255D\n");
        System.out.print(sb);
        System.out.flush();
    }

    private boolean esMejorMvp(Entidad candidato, Entidad actual) {
        if (candidato.getKills() != actual.getKills()) return candidato.getKills() > actual.getKills();
        return candidato.getDanioInfligido() > actual.getDanioInfligido();
    }

    private void lineaCentrada(StringBuilder sb, int ancho, String texto) {
        int visible = longitudVisible(texto);
        int pad = (ancho - visible) / 2;
        int padDer = ancho - visible - pad;
        sb.append("\u2551").append(espacios(pad)).append(texto).append(espacios(padDer)).append("\u2551\n");
    }

    private void lineaIzq(StringBuilder sb, int ancho, String texto) {
        int visible = longitudVisible(texto);
        sb.append("\u2551").append(texto).append(espacios(ancho - visible)).append("\u2551\n");
    }

    private int longitudVisible(String s) {
        return s.replaceAll("\u001B\\[[;\\d]*m", "").length();
    }

    private String espacios(int n) {
        return n > 0 ? " ".repeat(n) : "";
    }
}

File: web/js/renderer.js
import { Aliado, Enemigo, EnemigoTanque, EnemigoRapido, Muro } from './entidad.js';
import { Escudo, Arma, Estrella, Velocidad, Pocion, Trampa } from './objetos.js';

const COLORES_ESTRELLA = ['#ef4444', '#eab308', '#22c55e', '#06b6d4', '#a855f7', '#f5f5f5'];

export class Renderer {
    constructor(canvas, hudDiv, statsDiv) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        this.hudDiv = hudDiv;
        this.statsDiv = statsDiv;
    }

    drawBoard(board, turno) {
        const filas = board.filas;
        const columnas = board.columnas;
        const cellW = this.canvas.width / columnas;
        const cellH = this.canvas.height / filas;
        const ctx = this.ctx;

        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);

        const fontSize = Math.floor(Math.min(cellW, cellH) * 0.65);
        ctx.font = `bold ${fontSize}px monospace`;
        ctx.textAlign = 'center';
        ctx.textBaseline = 'middle';

        for (let f = 0; f < filas; f++) {
            for (let c = 0; c < columnas; c++) {
                const x = c * cellW;
                const y = f * cellH;
                const cx = x + cellW / 2;
                const cy = y + cellH / 2;

                const e = board.getEntidad(f, c);

                if (e !== null) {
                    if (e instanceof Muro) {
                        ctx.fillStyle = '#78716c';
                        ctx.fillRect(x, y, cellW, cellH);
                        ctx.fillStyle = '#eab308';
                        ctx.fillText('[=]', cx, cy);
                    } else if (e instanceof EnemigoTanque) {
                        ctx.fillStyle = '#1a1a2e';
                        ctx.fillRect(x, y, cellW, cellH);
                        ctx.fillStyle = '#dc2626';
                        ctx.fillText('T', cx, cy);
                    } else if (e instanceof EnemigoRapido) {
                        ctx.fillStyle = '#1a1a2e';
                        ctx.fillRect(x, y, cellW, cellH);
                        ctx.fillStyle = '#eab308';
                        ctx.fillText('\u00A4', cx, cy);
                    } else if (e instanceof Enemigo) {
                        ctx.fillStyle = '#1a1a2e';
                        ctx.fillRect(x, y, cellW, cellH);
                        ctx.fillStyle = '#ef4444';
                        ctx.fillText('#', cx, cy);
                    } else if (e instanceof Aliado) {
                        ctx.fillStyle = '#1a1a2e';
                        ctx.fillRect(x, y, cellW, cellH);
                        if (e.turnosInvencible > 0) {
                            ctx.fillStyle = COLORES_ESTRELLA[turno % COLORES_ESTRELLA.length];
                        } else {
                            ctx.fillStyle = '#22c55e';
                        }
                        ctx.fillText('o', cx, cy);
                    } else {
                        ctx.fillStyle = '#1a1a2e';
                        ctx.fillRect(x, y, cellW, cellH);
                    }
                } else if (board.getObjeto(f, c) !== null) {
                    ctx.fillStyle = '#1a1a2e';
                    ctx.fillRect(x, y, cellW, cellH);
                    const obj = board.getObjeto(f, c);
                    if (obj instanceof Escudo) {
                        ctx.fillStyle = '#06b6d4';
                        ctx.fillText('S', cx, cy);
                    } else if (obj instanceof Arma) {
                        ctx.fillStyle = '#a855f7';
                        ctx.fillText('W', cx, cy);
                    } else if (obj instanceof Velocidad) {
                        ctx.fillStyle = '#3b82f6';
                        ctx.fillText('V', cx, cy);
                    } else if (obj instanceof Estrella) {
                        ctx.fillStyle = '#facc15';
                        ctx.fillText('*', cx, cy);
                    } else if (obj instanceof Pocion) {
                        ctx.fillStyle = '#4ade80';
                        ctx.fillText('+', cx, cy);
                    } else {
                        ctx.fillStyle = '#ffffff';
                        ctx.fillText(obj.simbolo, cx, cy);
                    }
                } else if (board.getTrampa(f, c) !== null) {
                    ctx.fillStyle = '#1a1a2e';
                    ctx.fillRect(x, y, cellW, cellH);
                    ctx.fillStyle = '#9ca3af';
                    ctx.fillText('^', cx, cy);
                } else {
                    ctx.fillStyle = '#1a1a2e';
                    ctx.fillRect(x, y, cellW, cellH);
                    ctx.fillStyle = '#333347';
                    ctx.fillText('.', cx, cy);
                }
            }
        }

        // Grid lines
        ctx.strokeStyle = '#2a2a3e';
        ctx.lineWidth = 0.5;
        for (let f = 0; f <= filas; f++) {
            ctx.beginPath();
            ctx.moveTo(0, f * cellH);
            ctx.lineTo(this.canvas.width, f * cellH);
            ctx.stroke();
        }
        for (let c = 0; c <= columnas; c++) {
            ctx.beginPath();
            ctx.moveTo(c * cellW, 0);
            ctx.lineTo(c * cellW, this.canvas.height);
            ctx.stroke();
        }
    }

    updateHUD(engine) {
        const tiempoMs = Date.now() - engine.tiempoInicio;
        const seg = Math.floor(tiempoMs / 1000);
        const min = Math.floor(seg / 60);
        const s = seg % 60;
        const tiempo = `${String(min).padStart(2, '0')}:${String(s).padStart(2, '0')}`;

        const colorAliados = engine.numAliados <= 2 ? '#ef4444' : '#22c55e';

        this.hudDiv.innerHTML = `
            <div class="hud-row">
                <span>Turno: <strong>${engine.turno}</strong></span>
                <span>Tiempo: <strong>${tiempo}</strong></span>
                <span>Velocidad: <strong>${engine.config.velocidadMs}ms</strong></span>
            </div>
            <div class="hud-row">
                <span>Aliados: <strong style="color:${colorAliados}">${engine.numAliados}/${engine.config.numAliado}</strong></span>
                <span>Enemigos: <strong style="color:#ef4444">${engine.numEnemigos}</strong></span>
                <span>Eliminados: <strong style="color:#22c55e">${engine.enemigosEliminados}</strong></span>
                <span>Objetos: <strong style="color:#06b6d4">${engine.objetosRecogidos}</strong></span>
            </div>
            <div class="hud-leyenda">
                <span style="color:#22c55e">o</span>=Aliado
                <span style="color:#ef4444">#</span>=Enemigo
                <span style="color:#dc2626">T</span>=Tanque
                <span style="color:#eab308">\u00A4</span>=R\u00E1pido
                <span style="color:#eab308">[=]</span>=Muro
                <span style="color:#9ca3af">^</span>=Trampa
                <span style="color:#06b6d4">S</span>=Escudo
                <span style="color:#a855f7">W</span>=Arma
                <span style="color:#facc15">*</span>=Estrella
                <span style="color:#3b82f6">V</span>=Vel
                <span style="color:#4ade80">+</span>=Poci\u00F3n
            </div>
        `;
    }

    mostrarEstadisticas(engine) {
        let danioAliadosInf = 0, danioAliadosRec = 0, killsAliados = 0, objRecogidos = 0;
        let aliadosVivos = 0;
        let danioEnemigosInf = 0, danioEnemigosRec = 0, killsEnemigos = 0;
        let enemigosVivos = 0;
        let mvpAliado = null, mvpEnemigo = null;

        for (const e of engine.todasEntidades) {
            if (e instanceof Aliado) {
                danioAliadosInf += e.danioInfligido;
                danioAliadosRec += e.danioRecibido;
                killsAliados += e.kills;
                objRecogidos += e.objetosRecogidosPersonal;
                if (e.estaVivo()) aliadosVivos++;
                if (mvpAliado === null || this._esMejorMvp(e, mvpAliado)) mvpAliado = e;
            } else if (e instanceof Enemigo) {
                danioEnemigosInf += e.danioInfligido;
                danioEnemigosRec += e.danioRecibido;
                killsEnemigos += e.kills;
                if (e.estaVivo()) enemigosVivos++;
                if (mvpEnemigo === null || this._esMejorMvp(e, mvpEnemigo)) mvpEnemigo = e;
            }
        }

        const tiempoMs = Date.now() - engine.tiempoInicio;
        const seg = Math.floor(tiempoMs / 1000);
        const tiempo = `${String(Math.floor(seg / 60)).padStart(2, '0')}:${String(seg % 60).padStart(2, '0')}`;

        let textoResultado, colorResultado;
        switch (engine.resultado) {
            case "aliados":
                textoResultado = "Los aliados han ganado";
                colorResultado = "#22c55e";
                break;
            case "enemigos":
                textoResultado = "Los enemigos han ganado";
                colorResultado = "#ef4444";
                break;
            default:
                textoResultado = "Empate";
                colorResultado = "#eab308";
                break;
        }

        let mvpAliadoHTML = '';
        if (mvpAliado !== null) {
            mvpAliadoHTML = `
                <div class="stats-section-title" style="color:#22c55e">MVP ALIADO</div>
                <div class="stats-line">Aliado #${mvpAliado.id} \u2014 ${mvpAliado.kills} kills, ${mvpAliado.danioInfligido} da\u00F1o</div>
            `;
        }
        let mvpEnemigoHTML = '';
        if (mvpEnemigo !== null) {
            mvpEnemigoHTML = `
                <div class="stats-section-title" style="color:#ef4444">MVP ENEMIGO</div>
                <div class="stats-line">Enemigo #${mvpEnemigo.id} \u2014 ${mvpEnemigo.kills} kills, ${mvpEnemigo.danioInfligido} da\u00F1o</div>
            `;
        }

        this.statsDiv.innerHTML = `
            <div class="stats-header">ESTADISTICAS DE PARTIDA</div>
            <div class="stats-separator"></div>
            <div class="stats-line">Resultado: <strong style="color:${colorResultado}">${textoResultado}</strong></div>
            <div class="stats-line">Turnos: ${engine.turno} | Tiempo: ${tiempo}</div>
            <div class="stats-separator"></div>
            <div class="stats-section-title" style="color:#22c55e">ALIADOS</div>
            <div class="stats-line">Da\u00F1o infligido total: ${danioAliadosInf}</div>
            <div class="stats-line">Da\u00F1o recibido total: ${danioAliadosRec}</div>
            <div class="stats-line">Supervivientes: ${aliadosVivos}/${engine.numAliadosInicial}</div>
            <div class="stats-line">Kills totales: ${killsAliados}</div>
            <div class="stats-line">Objetos recogidos: ${objRecogidos}</div>
            <div class="stats-separator"></div>
            <div class="stats-section-title" style="color:#ef4444">ENEMIGOS</div>
            <div class="stats-line">Da\u00F1o infligido total: ${danioEnemigosInf}</div>
            <div class="stats-line">Da\u00F1o recibido total: ${danioEnemigosRec}</div>
            <div class="stats-line">Supervivientes: ${enemigosVivos}/${engine.numEnemigosInicial}</div>
            <div class="stats-line">Kills totales: ${killsEnemigos}</div>
            <div class="stats-separator"></div>
            ${mvpAliadoHTML}
            ${mvpEnemigoHTML}
        `;
        this.statsDiv.style.display = 'block';
    }

    _esMejorMvp(candidato, actual) {
        if (candidato.kills !== actual.kills) return candidato.kills > actual.kills;
        return candidato.danioInfligido > actual.danioInfligido;
    }
}
